File: cpp/0039组合总和.cpp
#include<vector>

using namespace std;

vector<vector<int>> res;
vector<int> temp;

void backTracking(int start,vector<int>&candidates,int target,int sum)
{
    if(sum == target)
    {
        res.push_back(temp);
        return;
    }

    for (int i = start;i<candidates.size()&&sum<=target;++i)
    {
        temp.push_back(candidates[i]);
        backTracking(i, candidates, target, sum + candidates[i]);
        temp.pop_back();
    }
}

vector<vector<int>> combinationSum(vector<int> &candidates, int target)
{
    backTracking(0, candidates, target, 0);
    return res;
}

File: python/main/0039组合总和.py

def combinationSum(candidates: list, target: int) -> list:
    res=[]
    temp=[]
    def _backtrakcing(candidates:list,target:int,s=0,st=0):

        if(s>target):
            return
        if(s==target):
            res.append(temp.copy())
            return
        

        for i in range(st,len(candidates)):
            temp.append(candidates[i])
            _backtrakcing(candidates,target,s+candidates[i],i)
            temp.pop()

    
    _backtrakcing(candidates,target)
    return res



print(combinationSum([2,3,6,7],7))